python/cudf_polars/cudf_polars/dsl/expressions/ternary.py

```diff
@@ -41,15 +42,33 @@ def do_evaluate(
         when, then, otherwise = (
             child.evaluate(df, context=context) for child in self.children
         )
-        then_obj = then.obj_scalar(stream=df.stream) if then.is_scalar else then.obj
-        otherwise_obj = (
-            otherwise.obj_scalar(stream=df.stream)
-            if otherwise.is_scalar
-            else otherwise.obj
+        if when.is_scalar:
+            # For scalar predicates: lowering to copy_if_else would require
+            # materializing an all true/false mask column. Instead, just pick
+            # the correct branch.
+            when_predicate = when.obj_scalar(stream=df.stream).to_py()
+            pick, other = (then, otherwise) if when_predicate else (otherwise, then)
+
+            if pick.is_scalar:
+                (pick_col,) = broadcast(
+                    pick,
+                    target_length=1 if other.is_scalar else other.size,
+                    stream=df.stream,
+                )
+                return Column(pick_col.obj, dtype=self.dtype)
+
+            return Column(pick.obj, dtype=self.dtype)
+
+        then_col, otherwise_col = broadcast(
+            then, otherwise, target_length=when.size, stream=df.stream
         )
+
```

Comment:
Please don't do this. If the then/otherwise columns are actually scalars libcudf will happily broadcast them out for us.

Reply:
> I think we want the logic to be the following:
> 
> ```python
> when_expr, then_expr, otherwise_expr = self.children
> when = when_expr.evaluate(...)
> if when.is_scalar:
>     pred = when.obj_scalar(stream=...).to_py(stream=...)
>     result, = broadcast(then_expr.evaluate(...) if pred else otherwise_expr.evaluate(...), target_length=df.num_rows, stream=df.stream)
> else:
>     then = then_expr.evaluate(...)
>     otherwise = otherwise_expr.evaluate(...)
>     then_obj = then.obj_scalar(stream=df.stream) if then.is_scalar else then.obj
>     otherwise_obj = ...
>     return Column(plc.copying.copy_if_else(then_obj, otherwise_obj, when.obj, stream=df.stream), ...)
> ```

slight correction: we don't want to broadcast to the dataframe's row count